scarlet/model.py

```diff
@@ -0,0 +1,118 @@
+from abc import ABC, abstractmethod
+
+from .parameter import Parameter
+
+
+class Model(ABC):
+    """Model base class.
+
+    This class stores and provides access to parameters and sub-ordinate models.
+
+    Parameters
+    ----------
+    parameters: list of `~scarlet.Parameter`
+    children: list of `~scarlet.Model`
+        Subordinate models.
+    """
+
+    def __init__(self, *parameters, children=None):
+
+        if len(parameters) == 0:
+            self._parameters = ()
+        elif isinstance(parameters, Parameter):
+            self._parameters = (parameters,)
+        elif isinstance(parameters, (list, tuple)):
+            for p in parameters:
+                assert isinstance(p, Parameter)
+            self._parameters = parameters
+        else:
+            raise TypeError(
+                "parameter must be None, a Parameter, or a list of Parameters"
+            )
+
+        if children is None:
+            children = ()
+        if hasattr(children, "__iter__"):
+            for c in children:
+                assert isinstance(c, Model)
+            self._children = children
+        else:
+            assert isinstance(children, Model)
+            self._children = tuple(children)
+
+        self.check_parameters()
+
+    @property
+    def parameters(self):
+        """List of parameters, including from the children.
+        """
+        return self._parameters + tuple(p for c in self.children for p in c.parameters)
+
+    @property
+    def children(self):
+        return self._children
+
+    def __getitem__(self, i):
+        return self._children.__getitem__(i)
+
+    def __iter__(self):
+        return self._children.__iter__()
+
+    def __next__(self):
+        return self._children.__next__()
+
+    def get_parameter(self, i, *parameters):
+        # NOTE: index lookup only works if order is not changed by parameter fixing!
+        # check parameters first during optimization
+        if i < len(parameters):
+            return parameters[i]
+
+        # eitherwise use self
+        if i < len(self.parameters):
+            return self.parameters[i]
+
+        return None
+
+    @abstractmethod
+    def get_model(self, *parameters, **kwargs):
+        """Get the model realization
+
+        Parameters
+        ----------
+        parameters: tuple of optimimzation parameters
+
+        Returns
+        -------
+        model: array
+            Realization of the model
+        """
+        pass
+
+    def get_models_of_children(self, *parameters, **kwargs):
```

Comment:
I like the idea, but that's quite a mouthful. Maybe rename to `get_child_models`?

Reply:
Grammar! It's not several models for a single child. I'm OK with `get_children_models`, which is exactly 3 chars shorter.

Reply:
I don't think that is grammatically correct, or at least if it is it's awkward. Each child _is_ a model, so we want to get all of the _child models_. You read it as "get all of this child's models." But if you don't like that we can just leave it as is.